libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -142,6 +142,11 @@ void CompilerUtils::storeInMemory(unsigned _offset)
 
 void CompilerUtils::storeInMemoryDynamic(Type const& _type, bool _padToWordBoundaries)
 {
+	//@todo verify if its of benefit/low-effort to provide Type::canBeStoredInMemory()
+	solUnimplementedAssert(_type.canBeStored() && _type.isValueType(),
```

Comment:
I actually don't think anymore we can use an assert here.

Reply:
There is something wrong with all of this, possibly even the filed issue, but not sure. Just let me a little bit more time (remember, I'll do this in parallel with teach-in).

Reply:
Exactly, the issue is wrong.

Reply:
The "should-never-happen" case I see is when "numBytes" is 0:

```cpp
		unsigned numBytes = prepareMemoryStore(_type, _padToWordBoundaries);
		if (numBytes > 0)
		{
			solUnimplementedAssert(
				_type.sizeOnStack() == 1,
				"Memory store of types with stack size != 1 not implemented."
			);
			m_context << Instruction::DUP2 << Instruction::MSTORE;
			m_context << u256(numBytes) << Instruction::ADD;
		}
```

nothing happens

so possibly an `else` is needed, with an assert

```cpp
		if (numBytes > 0)
		{
			solUnimplementedAssert(
				_type.sizeOnStack() == 1,
				"Memory store of types with stack size != 1 not implemented."
			);
			m_context << Instruction::DUP2 << Instruction::MSTORE;
			m_context << u256(numBytes) << Instruction::ADD;
		}
		else
		{
			solUnimplementedAssert(false,
				"Memory store of type with stack size 0"
			);
		}
```

which would then become

```cpp
		solAssert(numBytes > 0,
			"Memory store of type with stack size 0"
		);

		solUnimplementedAssert(
			_type.sizeOnStack() == 1,
			"Memory store of types with stack size != 1 not implemented."
		);
		m_context << Instruction::DUP2 << Instruction::MSTORE;
		m_context << u256(numBytes) << Instruction::ADD;
```

Reply:
I still want to give it a try (its anyway good for teachin).